File: JAVASCRIPT/Cap08/c08ex07.js
// c08ex07.js

const rls = require('readline-sync');

const LIMITE = 10;

let MATRIZ = new Array(LIMITE).fill(0);
let TOPO = 0;

function Vazia() {
  return TOPO == 0;
}

function Cheia() {
  return TOPO == LIMITE;
}

function Adicionar(ELEMENTO) {
  if (Cheia())
    return false;
  else {
    ++TOPO;
    MATRIZ[TOPO] = ELEMENTO;
    return true;
  }
}

function Retirar() {
  if (Vazia())
    return null;
  else {
    const ELEMENTO = MATRIZ[TOPO];
    MATRIZ[TOPO] = 0;
    --TOPO;
    return ELEMENTO;
  }
}

function Empilhar() {
  const X = rls.questionInt("Entre com um elemento numerico: ");
  if (Adicionar(X))
    console.log(`\nElemento ${X} inserido na posicao ${TOPO}.`);
  else
    console.log(`\nImpossivel adicionar ${X} - pilha cheia.`);
  console.log("");
}

function Desempilhar() {
  const ELEMENTO = Retirar();
  if (ELEMENTO !== null)
    console.log(`Elemento ${ELEMENTO} retirado do topo da pilha.`);
  else
    console.log("Impossivel retirar elemento - pilha vazia.");
  console.log("");
}

function Mostrar() {
  if (!Vazia())
    for (let I = TOPO; I >= 1; --I)
      console.log(`Posicao: ${I} = ${MATRIZ[I]}`);
  else
    console.log("Impossivel apresentar - pilha vazia.");
  console.log("");
}

function Criar() {
  TOPO = 0;
  for (let I = 0; I <= LIMITE - 1; ++I)
    MATRIZ[I] = 0;
}

let Opcao = 0;

Criar();
Opcao = 0;
while (Opcao !== 5) {
  console.log("PROGRAMA: PILHA\n");
  console.log("[1] - Empilhar");
  console.log("[2] - Desempilhar");
  console.log("[3] - Mostrar");
  console.log("[4] - Criar pilha");
  console.log("[5] - Sair\n");
  Opcao = rls.questionInt("Escolha uma opcao: ");
  console.log("");
  if (Opcao != 5) {  
    switch (Opcao) {
      case 1: Empilhar();    break;
      case 2: Desempilhar(); break;
      case 3: Mostrar();     break;
      case 4: Criar();       break;
      default:
        console.log("Opcao invalida. Tente novamente.\n");
    }
  }
}

File: SWIFT/Cap08/c08ex07.swift
// c08ex07.swift

import Foundation

let LIMITE = 10
var MATRIZ = [Int](repeating: 0, count: LIMITE)
var TOPO = -1

func Vazia() -> Bool {
    if (TOPO == -1) {
      return true
    } else {
      return false
    }
}

func Cheia() -> Bool {
    if (TOPO == LIMITE - 1) {
      return true;
    } else {
      return false;
    }
}

func Adicionar(_ ELEMENTO: Int) -> Bool {
    if (Cheia()) {
        return false
    } else {
        TOPO += 1
        MATRIZ[TOPO] = ELEMENTO
        return true
    }
}

func Retirar(_ ELEMENTO: inout Int) -> Bool {
    if (Vazia()) {
        return false
    } else {
        ELEMENTO = MATRIZ[TOPO]
        MATRIZ[TOPO] = 0
        TOPO -= 1
        return true
    }
}

func Empilhar() {
    print("Entre com um elemento numerico: ", terminator: "")
    if let X = Int(readLine() ?? "") {
        if (Adicionar(X)) {
            print("\nElemento \(X) inserido na posicao \(TOPO).\n")
        } else {
            print("\nImpossivel adicionar \(X) - pilha cheia.\n")
        }
    } else {
        print("\nEntrada invalida.\n")
    }
}

func Desempilhar() {
    var ELEMENTO: Int = 0
    if (Retirar(&ELEMENTO)) {
        print("Elemento \(ELEMENTO) retirado do topo da pilha.\n")
    } else {
        print("Impossivel retirar elemento - pilha vazia.\n")
    }
}

func Mostrar() {
    if (!(Vazia())) {
        for i in (0...TOPO).reversed() {
            print("Posicao: \(String(format: "%2d", i+1)) = \(MATRIZ[i])")
        }
    } else {
        print("Impossivel apresentar - pilha vazia.\n")
    }
    print()
}

func Criar() {
    TOPO = -1
    for i in 0...LIMITE - 1 {
        MATRIZ[i] = 0
    }
}

var opcao = 0
while (opcao != 5) {
    print("PROGRAMA: PILHA\n")
    print("[1] - Empilhar")
    print("[2] - Desempilhar")
    print("[3] - Mostrar")
    print("[4] - Criar pilha")
    print("[5] - Sair\n")
    print("Escolha uma opcao: ", terminator: "")

    if let entrada = readLine(), let op = Int(entrada) {
        opcao = op
        print()
        if (opcao != 5) {
            switch opcao {
                case 1: Empilhar()
                case 2: Desempilhar()
                case 3: Mostrar()
                case 4: Criar()
                default: print("Opcao invalida. Tente novamente.\n")
            }
        }
    } else {
        print("Entrada invalida. Tente novamente.\n")
    }
}
